cmd/maya-exporter/app/collector/jivacollector.go

```diff
@@ -0,0 +1,216 @@
+	j.Metrics.connectionRetryCounter.Collect(ch)
+	j.Metrics.connectionErrorCounter.Collect(ch)
+}
+
+// collector selects the container attached storage for the collection of
+// metrics.Supported CAS are jiva and cstor.
+func (j *JivaStatsExporter) collector() error {
+	// collect the metrics from jiva controller and send it via channels
+	if err := j.collect(); err != nil {
+		j.Metrics.connectionErrorCounter.WithLabelValues(err.Error()).Inc()
+		glog.Error("Error in collecting metrics, found error:", err)
+		return errors.New("error in collecting metrics")
+	}
+	return nil
+}
+
+// NewJivaStatsExporter returns Jiva volume controller URL along with Path.
+func NewJivaStatsExporter(volumeControllerURL *url.URL) *JivaStatsExporter {
+	volumeControllerURL.Path = v1.StatsAPI
+	return &JivaStatsExporter{
+		VolumeControllerURL: volumeControllerURL.String(),
+		Metrics:             metrics,
+	}
+}
+
+// getVolumeStats is used to decode the response from the Jiva controller
+// response received by the client is in json format which then decoded
+// and mapped to VolumeMetrics.
+func (j *JivaStatsExporter) getVolumeStats(obj interface{}) error {
+	httpClient := http.DefaultClient
+	httpClient.Timeout = 1 * time.Second
+	resp, err := httpClient.Get(j.VolumeControllerURL)
+
+	if err != nil {
+		glog.Error("could not retrieve OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(obj)
+
+	if err != nil {
+		glog.Error("could not decode OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	defer resp.Body.Close()
+	return nil
+}
+
+// collect is used to set the values gathered from OpenEBS volume controller
+// to prometheus gauges.
+func (j *JivaStatsExporter) collect() error {
+	var (
+		initialMetrics, finalMetrics v1.VolumeMetrics
+		m                            MetricsDiff
+	)
+
+	err := j.getVolumeStats(&initialMetrics)
+	if err != nil {
+		glog.Error("Could not decode: %v", err)
+		return err
+	}
+
+	time.Sleep(1 * time.Second)
+	err = j.getVolumeStats(&finalMetrics)
+	if err != nil {
+		glog.Error("Could not decode: %v", err)
+		return err
+	}
+
+	m = j.parser(initialMetrics, finalMetrics)
+
+	j.Metrics.readIOPS.Set(m.readIOPS)
+	j.Metrics.readTimePS.Set(m.readTimePS)
+	j.Metrics.readBlockCountPS.Set(m.readBlockCountPS)
+	j.Metrics.writeIOPS.Set(m.writeIOPS)
+	j.Metrics.writeTimePS.Set(m.writeTimePS)
+	j.Metrics.writeBlockCountPS.Set(m.writeBlockCountPS)
+	j.Metrics.sectorSize.Set(m.sectorSize)
+	j.Metrics.logicalSize.Set(m.logicalSize)
+	j.Metrics.actualUsed.Set(m.actualSize)
+	j.Metrics.sizeOfVolume.Set(m.size)
+	url := j.VolumeControllerURL
+	url = strings.TrimSuffix(url, ":9501/v1/stats")
+	url = strings.TrimPrefix(url, "http://")
+	j.Metrics.volumeUpTime.WithLabelValues(finalMetrics.Name, "iqn.2016-09.com.openebs.jiva:"+finalMetrics.Name, url).Set(finalMetrics.UpTime)
+	return nil
+}
+
+func (j *JivaStatsExporter) parser(m1, m2 v1.VolumeMetrics) MetricsDiff {
```

Comment:
same comments as earlier. please check if parser can be a method of Metrics itself.
In other words, can be move this entire piece to Metrics. It will do the Parse stuff based on the registrations & respond with a MetricsDiff.
Query - Is Metrics common across JivaExporter & CStorExporter. It will be great if it is common.

Reply:
yes